roles/jd-server/src/lib/job_declarator/mod.rs

```diff
@@ -450,39 +453,77 @@ impl JobDeclarator {
             if let Ok((receiver, sender, _, _)) =
                 Connection::new(stream, HandshakeRole::Responder(responder)).await
             {
-                let setup_message_from_proxy_jd = receiver.recv().await.unwrap();
-                info!(
-                    "Setup connection message from proxy: {:?}",
-                    setup_message_from_proxy_jd
-                );
-
-                let setup_connection_success_to_proxy = SetupConnectionSuccess {
-                    used_version: 2,
-                    // Setup flags for async_mining_allowed
-                    flags: 0b_0000_0000_0000_0000_0000_0000_0000_0001,
+                let mut incoming = match receiver.recv().await {
+                    Ok(EitherFrame::Sv2(s)) => {
+                        debug!("Got sv2 message: {:?}", s);
+                        s
+                    }
+                    Ok(EitherFrame::HandShake(s)) => {
+                        error!(
+                            "Received unexpected handshake message from upstream: {:?} at {:?}",
+                            s, addr
+                        );
+                        continue;
+                    }
+                    Err(e) => {
+                        error!("Error receiving message: {:?}", e);
+                        continue;
+                    }
                 };
-                let sv2_frame: StdFrame =
-                    JdsMessages::Common(setup_connection_success_to_proxy.into())
-                        .try_into()
-                        .unwrap();
-                let sv2_frame = sv2_frame.into();
-                info!("Sending success message for proxy");
-                sender.send(sv2_frame).await.unwrap();
 
-                let jddownstream = Arc::new(Mutex::new(JobDeclaratorDownstream::new(
-                    receiver.clone(),
-                    sender.clone(),
-                    &config,
-                    mempool.clone(),
-                    // each downstream has its own sender (multi producer single consumer)
-                    sender_add_txs_to_mempool.clone(),
-                )));
+                let payload = incoming.payload();
+
+                if let Ok(value) = binary_sv2::from_bytes::<SetupConnection>(payload) {
+                    let flag = value.flags;
+                    let is_valid =
+                        SetupConnection::check_flags(Protocol::JobDeclarationProtocol, flag, 1);
```

Comment:
here the JDS offers async job mining, so if the downstream does not requires it, the flags checking should produce `True`.
Aslo, according to the fact that the client set or not the flag `async mining job`, we espect different behaviours from the JDS. Perhaps this is material for a second PR.

Reply:
you are right!

and I think we should fix that on this PR, otherwise this PR would be introducing a broken feature into JDS

I elaborate more here https://github.com/stratum-mining/stratum/pull/1035#issuecomment-2237516417